### DISCUSSION TITLE
Flask Socketio not emitting data to client after running "with app.app_context()" on login process

### DISCUSSION THREAD

**Discussion by zaingithub** [miguelgrinberg/Flask-SocketIO#1794] [General]
Hello, I'm developing a desktop application using flask.

Where in this application users can login using their email and password.

While using flask-soscketio, I encountered problems logging in.

When the user logs in, the application will authenticate the email and password, if it is successful then my application will call a function to run the scheduler using flask-apscheduler (https://github.com/viniciuschiele/flask-apscheduler) to check the license periodically.

The code is roughly like this:

```
def task_license_scheduler() :
	with app.app_context(), app.test_request_context() :
		scheduler.add_job(code_here)

@auth.route('/login', methods=['GET', 'POST'])
def login():
	if request.method == 'POST':
		is user_valid :
			login_user(user_valid)
			### run scheduler
			task_license_scheduler()
```

When the user has successfully logged in, they will be directed to the "homepage", the server should have emitted data to the client.

But when the application calls the **task_license_scheduler()** function, the server emits any data to the client.

But when I restart the flask application, the socket runs normally and emits data to the client as expected.

This happens only when inside the **task_license_scheduler()** function involves **with app.app_context(), app.test_request_context()**.

Please give the best solution for this problem. thank you very much.

**Comment by miguelgrinberg** [comment]
I'm sorry but I don't understand what the problem is. In any case, using a test request context in an application makes absolutely no sense. Not sure what you are trying to achieve, but a test request is only for testing.

**Comment by zaingithub** [comment]
**bot.py**

```
from app import create_app, socketio

app = create_app(debug=True)

if __name__ == '__main__':
       socketio.run(app)
```

router blueprint :

**auth.py**

```
def cron_verify_license():
       from bot import app ### imported app flask from bot.py
       with app.app_context(), app.test_request_context() :
              run_license_checker()

def task_license_scheduler() :
       job_id = md5('check_license_{}'.format(Loader.deviceid))
       if not scheduler.get_job(job_id) :
              scheduler.add_job(
                     func=cron_verify_license,
                    args=[ job_id ],
                    trigger='interval',
                    seconds=120,
                    replace_existing=True,
                    id=job_id
             )

@auth.route('/login', methods=['GET', 'POST'])
def login():
       if request.method == 'POST':
              user_valid = check_user_from_db()
              if user_valid :
                    login_user(user_valid)
                    ### run scheduler
                    task_license_scheduler()
```

**Reply by miguelgrinberg** [reply]
There is no single `emit()` in this code, so this code is incomplete. Also, as I mentioned before, you can't use a test request context in an application, that is not what test request contexts were designed for.

**Reply by zaingithub** [reply]
router **events.py**

```
from .. import socketio

@socketio.on('my_event')
def my_event(message):
	socketio.emit('my_response', {'data': 'got it!'})

@socketio.on('connected_event')
def connected(msg):
	socketio.emit('server_response', {'data': msg['data']})

@socketio.on('broadcast_event')
def broadcast(msg):
	socketio.emit('server_response', {'data': msg['data']}, broadcast=True)
```


template **index.html**

```
<!DOCTYPE html>
<html>
<head>
	<title>Test</title>
</head>
<body>

	<div id="app">
	    <div class="container">
	        <div class="row">
	            <div class="col-sm-5">
	                <input class="form-control" type="text" v-model="clientMsg" 
	                       @keyup.enter="emitMsg"/>
	            </div>
	            <div class="col-sm-2">
	                <button class="btn btn-success" id="emitMsg" @click="emitMsg">Send</button>
	            </div>
	            <div class="col-sm-5">
	            </div>
	        </div>
	    </div>

	    <br />
	    <pre>@{ log }</pre>

	    <hr>

	</div>

	<script src="https://cdn.jsdelivr.net/npm/vue/dist/vue.js"></script>
	<script src="https://cdn.jsdelivr.net/npm/socket.io-client/dist/socket.io.slim.js"></script>
	<script src="https://cdn.jsdelivr.net/npm/vue-socket.io-extended"></script>


	  <script type="text/javascript">
	    var socket = io('http://localhost:1111');
	    Vue.use(VueSocketIOExt, socket);
	  </script>

	  <script type="text/javascript">

	    new Vue({
	        delimiters:['@{', '}'],
	        el: "#app",
	        sockets: {
	            connect: function() {
	                this.appendLog("Start connect...");
	                this.$socket.client.emit('connected_event', { data: 'WebSocket connected successfully.' });
	            },
	            connection: function(msg){
	            	alert(msg.data);
	            },
	            server_response: function(msg) {
	                this.appendLog(msg.data);
	            }
	        },
	        data: {
	            clientMsg: "",
	            log: "",
	        },
	        methods: {
	            emitMsg() {
	                this.$socket.client.emit('broadcast_event', { data: this.clientMsg });
	            },
	            appendLog(newLog) {
	                this.log += newLog + "\n";
	            }
	        },
	        created: function () {
	        }
	    })
	  </script>



</body>
</html>
```

**Preview**

![image](https://user-images.githubusercontent.com/23699888/156167309-3deb13e1-2e88-4751-be52-ee03f93a1e4b.png)

**Reply by zaingithub** [reply]
The *"send"* button on the template will send input data to the server, then the server will emit the input data to the client.

basically, before the *cron_verify_license* function runs, the client can still send data to the server by pressing the send button.

But after the *cron_verify_license* function runs for the first time, then the send button doesn't give any response to the server or emits it to the client.

**Reply by zaingithub** [reply]
is there any suggestion for me to be able to run this scheduler without affecting socketio?

**Reply by zaingithub** [reply]
![image](https://user-images.githubusercontent.com/23699888/156172919-09f79ccf-ec52-4fe0-925c-14652e51dd73.png)


this is my socketio log,

first I input "12345", I send, the server emits data "12345" to the client and displays it in html

second, I input "123456", I send, the client sends data to the server, but the server does not emit data to the client like the first.

between the first and second step, that's when my *cron_verify_license* function runs.

**Reply by miguelgrinberg** [reply]
You are using eventlet, which is likely incompatible with appscheduler. Try removing eventlet and see if that helps.

**Comment by zaingithub** [comment]
That is my program code...

When the user has successfully logged in, the bot will add a scheduler by running **task_license_scheduler()**.
The scheduler will later run the **cron_verify_license** function


Because there is an interval of 120 seconds, 
when the user opens the **home** page before the **cron_verify_license** function runs, socketio runs smoothly.
the server still can send (emit) data to the client.


but when the function **cron_verify_license** runs for the first time (in the app context like in my code), 
after that socketio doesn't run anymore.

I checked, when I called this code in my function : 

```
from bot import app
```

or

```
from flask import current_app
```

both will stop socketio to emit data to the client.

I don't know how to solve this problem.

While the **cron_verify_license** function I have to run periodically.

**Comment by zaingithub** [comment]
@miguelgrinberg 

![image](https://user-images.githubusercontent.com/23699888/156172919-09f79ccf-ec52-4fe0-925c-14652e51dd73.png)


this is my socketio log,

first I input "12345", I send, the server emits data "12345" to the client and displays it in html

second, I input "123456", I send, the client sends data to the server, but the server does not emit data to the client like the first.

between the first and second step, that's when my *cron_verify_license* function runs.

**Comment by zaingithub** [comment]
### SOLVED : Replace flask-socketio with python-socketio

----

![image](https://user-images.githubusercontent.com/23699888/156185452-5a7544c2-57c6-416e-b7f9-f9678c36cecb.png)

Hello, @miguelgrinberg, I tried to **replace flask-socketio with python-socketio** for my application, and everything went smoothly as I expected.

So I still believe this might be a **bug** of flask_socketio, and I'm still very curious about this. haha...

These are some steps I did to replace **flask_socketio to python-socketio**

----

I changed in my file **app/_ _init_ _.py** :

```
from flask_socketio import SocketIO
socketio 		= SocketIO(
        cors_allowed_origins='*', 
 	async_mode='gevent',
        logger=True,
        engineio_logger=True
 )
```

Becomes

```
import socketio
from engineio.async_drivers import gevent ### for pyinstaller
socket_io = socketio.Server(async_mode='gevent')
```

file **app/_ _init_ _.py** :

**_without socketio.init_app(app)_**

```

def create_app(debug=False):
	app 			= Flask(__name__, static_folder=storage_path("static"))
	app.wsgi_app 	= socketio.WSGIApp(socket_io, app.wsgi_app)
	....
	....
        #socketio.init_app(app)
	return app
```

file **bot.py**

```
from gevent import monkey
monkey.patch_all()
from gevent.pywsgi import WSGIServer
from geventwebsocket.handler import WebSocketHandler
from app import create_app

app = create_app(debug=True)

if __name__ == '__main__':
	WSGIServer(('127.0.0.1', 1111), app,
	                  handler_class=WebSocketHandler).serve_forever()
```

file **events.py**

the router code on the flask blueprint, as well as the socketio emit is more or less the same.

there is only an adjustment for adding the **sid** argument to the socketio event function like this

```
@socket_io.on('broadcast_event')
def broadcast(sid, msg):
	print("broadcast_event", msg)
	socket_io.emit('server_response', {'data': msg['data']})
```

**Reply by miguelgrinberg** [reply]
Okay, but this basically removes the integration between Flask and Socket.IO. Maybe that is okay for your use case, but not really a great idea, since now you will not be able to share session data between the two sides.

**Comment by zaingithub** [comment]
Yes I know that this is not the best solution at the moment. But I think this is enough, because I will bundle my application into an exe using pyinstaller. So the application will only be run by the user on the computer only. 

My goal to use socketio is only to update data in real time into my vue js template, so that the client doesn't have to refresh or call ajax periodically to get updated data from the server (backend)

But I still wish I could use flask_socketio again.

Maybe if you find a solution to what I call this as bug, i will go back to using flask_socketio again.

Thank you
